### ISSUE TITLE
Opinions on constexpr being added to C2X?

### ISSUE THREAD

**Issue by MarcusJohnson91** [rui314/chibicc#73]
There's a new proposal to add it, and the standard is especially interested in hearing from small compiler devs their opinions on such a feature.

http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2851.pdf

**Comment by rui314** [comment]
It looks like implementing the feature to a small compiler shouldn't be too hard, so I'd implement it to chibicc if it is accepted as a part of C2X. That being said, I'm not totally convinced that we _need_ it. One of the virtue of the C language is its stability. The proposed feature may be sometimes useful, but we can live without it.

**Comment by MarcusJohnson91** [comment]
I actually want it to have compile time test registration, i guess it can technically be done with a ton of macro magic and linker tricks, but that's over my head.

Thanks for the response btw, I'll link this issue to the WG14 guys

**Comment by rui314** [comment]
How can you use constexpr to register test functions at compile time? I'm just curious.

**Comment by MarcusJohnson91** [comment]
Call a constexpr function to add the current test to a global array of tests or a test suite struct with an array of test cases.

**Comment by rui314** [comment]
Can a constexpr function have a side effect such as adding a new element to a global array?

**Comment by MarcusJohnson91** [comment]
Yes, the proposal I linked at the beginning mentions that the constexpr function can modify it's own arguments.

 I don't see why passing the global testsuite/array as a parameter would violate that constraint.
 
 It's not like it's modifying errno or something like that, AFAICT it's basically just a regular function that happens to run at compile time.
 
 and the fact that it's at compiletime, you wouldn't even need to create like a VLA, the compiler could size it correctly right then and there and make that array fixed size (and constant)

**Comment by rui314** [comment]
So, is the following code valid?

```
constexpr int fn(int *arr) {
  return (arr[0]++, 5);
}

constexpr int g1[3]; // initialized with {0, 0, 0}?
int g2 = fn(g1);
```

`++` has a side effect, but as a whole the constexpr function consists of a single returns tatement.

**Comment by rui314** [comment]
One more question: can you take an address of a constexpr object at compile time and do pointer arithmetic on that pointer? I guess the answer is yes, because otherwise you can't really pass an array (or more precisely, the pointer to the first element of an array) to a constexpr function and access its elements.

**Comment by MarcusJohnson91** [comment]
> can you take an address of a constexpr object at compile time and do pointer arithmetic on that pointer?

Yes, he explicitly mentions in the paper, it's like 4 pages long.

**Comment by rui314** [comment]
I carefully read the proposed changes to the wording of the standard, and it isn't clear to me what is allowed, so I asked the questions. I'm sorry if the questions sounded silly. But please allow me to leave one comment and another question:

1. It says that a constexpr function is only allowed to modify or observe a state of its own argument or a return value. One interpretation of that is that a constexpr which takes a pointer can observe the value of the pointer itself but can't dereference  it, as the value pointed by a pointer isn't a state of the given argument itself.

2. If a constexpr function is allowed to mutate a given argument, what is the evaluation order if there are more than one top-level statements? I assume it is top to bottom, but it looks like the proposed changes don't specify that.

**Comment by MarcusJohnson91** [comment]
No no no, feel free to ask questions.

I'm not the author of the paper, I'm just a fan of getting constexpr into C.

I don't know the answer to these questions, I just invited the author of the proposal to drop by and answer your questions though :)

**Comment by GabrielRavier** [comment]
Although I am not the author of the paper, I think I can still answer this:

> I carefully read the proposed changes to the wording of the standard, and it isn't clear to me what is allowed, so I asked the questions. I'm sorry if the questions sounded silly. But please allow me to leave one comment and another question:
> 
> 1. It says that a constexpr function is only allowed to modify or observe a state of its own argument or a return value. One interpretation of that is that a constexpr which takes a pointer can observe the value of the pointer itself but can't dereference it, as the value pointed by a pointer isn't a state of the given argument itself.

This is indeed the case, although the statement that "A constexpr function does not modify or observe any state outside of its own arguments and return value" is more of a statement of fact than anything else as assignments or dereferences are not constant expressions and can thus not be used in a constexpr function.

> 2. If a constexpr function is allowed to mutate a given argument, what is the evaluation order if there are more than one top-level statements? I assume it is top to bottom, but it looks like the proposed changes don't specify that.

I'm not sure exactly what you mean by this, but the standard specifies that a `constexpr` function can only contain a single return statement evaluating a constant expression, so considering that fact I think the question is moot ?
